src/hotspot/share/memory/memoryReserver.cpp

```diff
@@ -433,13 +433,13 @@ ReservedSpace HeapReserver::Instance::try_reserve_range(char *highest_start,
   const uint64_t num_attempts_to_try   = MIN2((uint64_t)HeapSearchSteps, num_attempts_possible);
 
   const size_t stepsize = (attach_range == 0) ? // Only one try.
-    (size_t) highest_start : align_up(attach_range / num_attempts_to_try, attach_point_alignment);
+    1 : align_up(attach_range / num_attempts_to_try, attach_point_alignment);
 
-  // Try attach points from top to bottom.
-  for (char* attach_point = highest_start;
-       attach_point >= lowest_start && attach_point <= highest_start;  // Avoid wrap around.
-       attach_point -= stepsize) {
-    ReservedSpace reserved = try_reserve_memory(size, alignment, page_size, attach_point);
+  // Try reserve memory from top to bottom.
+  for (size_t offset = attach_range;
+       offset <= attach_range; // Avoid wrap around.
+       offset -= stepsize) {
+    ReservedSpace reserved = try_reserve_memory(size, alignment, page_size, lowest_start + offset);
```

Comment:
I think I preferred the earlier version over this.  But I don't dislike this version enough to want to block it.
I liked the explicitness of the pointer-based iteration path, and am unbothered by the 2nd loop exit.

Reply:
I have update the code for use the explicitness of the pointer-based iteration path, and convert `attach_point` to size_t and then convert to `char*` again, to avoid ubsan report "produced null pointer" fail.